File: Server/src/main/java/com/github/dr/rwserver/util/file/LoadConfig.java
package com.github.dr.rwserver.util.file;

import com.alibaba.fastjson.JSONObject;
import com.alibaba.fastjson.serializer.SerializerFeature;
import com.github.dr.rwserver.data.json.Json;
import com.github.dr.rwserver.struct.OrderedMap;
import com.github.dr.rwserver.util.log.Log;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

import static com.github.dr.rwserver.util.IsUtil.isBlank;

/**
 * @author Dr
 */
public final class LoadConfig {


    private final OrderedMap<String, String> data = new OrderedMap<>();
    private final FileUtil fileUtil;

    public LoadConfig(String file,boolean isFile) {
        fileUtil = isFile ? FileUtil.getFolder(file) : new FileUtil(file);
        reLoadConfig();
    }

    public LoadConfig(String file,String name) {
        fileUtil = FileUtil.getFolder(file).toFile(name);
        reLoadConfig();
    }

    public void reLoadConfig() {
        if (fileUtil.notExists() || fileUtil.readFileStringData().isEmpty()) {
            Log.error("NO Config.Json Use default configuration");
            return;
        }
        Json json = new Json(fileUtil.readFileStringData());
        //json对象转Map
        json.getInnerMap().forEach((k,v) -> data.put(k,v.toString()));
    }

    public void setConfig(String k,String v) {
       this.data.put(k,v);
    }
	private String load(String input, Object def) {
        String result = data.get(input);
        if (result == null) {
            Log.clog("NO KEY- Please check the file",input);
            data.put(input,def.toString());
            return def.toString();
        }
        return result;
	}

    public String readString(String input) {
        return readString(input,"");
    }

    public String readString(String input,@NotNull Object def) {
        String str = load(input,def);
        return isBlank(str) ? "" : str;
    }

    public int readInt(String input,@NotNull Object def) {
        String str = load(input,def);
        return Integer.parseInt(str);
    }

    public boolean readBoolean(String input,@NotNull Object def) {
        String str = load(input,def);
        return Boolean.parseBoolean(str);
    }

    public float readFloat(String input,@NotNull Object def) {
        String str = load(input,def);
        return Float.parseFloat(str);
    }

    public long readLong(String input,@NotNull Object def) {
        String str = load(input,def);
        return Long.parseLong(str);
    }

    public void setObject(String input,@NotNull Object key) {
        data.put(input,key.toString());
    }

    public void save() {
        final Map<String,String> map = new HashMap<>();
        data.each(map::put);
        fileUtil.writeFile(JSONObject.toJSONString(map, SerializerFeature.PrettyFormat),false);
        Log.clog("SAVE CONFIG OK");
    }
    public OrderedMap<String, String> getData() {
        return data;
    }
}

File: Server/src/main/java/cn/rwhps/server/util/file/LoadConfig.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package cn.rwhps.server.util.file

import cn.rwhps.server.data.json.Json
import cn.rwhps.server.data.json.Json.Companion.toJson
import cn.rwhps.server.struct.OrderedMap
import cn.rwhps.server.util.IsUtil.isBlank
import cn.rwhps.server.util.file.FileUtil.Companion.getFolder
import cn.rwhps.server.util.log.Log.clog
import cn.rwhps.server.util.log.Log.error

/**
 * @author Dr
 */
class LoadConfig {
    private val data = OrderedMap<String, Any>()
    private val fileUtil: FileUtil

    constructor(file: String, isFile: Boolean) {
        fileUtil = if (isFile) getFolder(file) else FileUtil(file)
        reLoadConfig()
    }

    constructor(file: String, name: String) {
        fileUtil = getFolder(file).toFile(name)
        reLoadConfig()
    }

    fun reLoadConfig() {
        if (fileUtil.notExists() || fileUtil.readFileStringData().isEmpty()) {
            error("NO Config.Json Use default configuration")
            return
        }
        val json = Json(fileUtil.readFileStringData())
        //json对象转Map
        json.getInnerMap().forEach { (key: String, value: Any) -> data.put(key, value) }
    }

    private fun load(input: String, def: Any): String {
        val result = data[input]
        if (result == null) {
            clog("NO KEY- Please check the file", input)
            data.put(input, def)
            return def.toString()
        }
        return result.toString()
    }

    @JvmOverloads
    fun readString(input: String, def: Any = ""): String {
        val str = load(input, def)
        return if (isBlank(str)) "" else str
    }

    fun readInt(input: String, def: Int): Int {
        val str = load(input, def)
        return str.toInt()
    }

    fun readBoolean(input: String, def: Boolean): Boolean {
        val str = load(input, def)
        return str.toBoolean()
    }

    fun readFloat(input: String, def: Float): Float {
        val str = load(input, def)
        return str.toFloat()
    }

    fun readLong(input: String, def: Long): Long {
        val str = load(input, def)
        return str.toLong()
    }

    fun setObject(input: String, key: Any) {
        data.put(input, key)
    }

    fun save() {
        val map: MutableMap<String, Any> = HashMap()
        data.each { key: String, value: Any -> map[key] = value }
        fileUtil.writeFile(toJson(map), false)
        clog("SAVE CONFIG OK")
    }
}